iceoryx_examples/automotive_soa/include/owl/kom/event_publisher_uds.inl

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_EXAMPLES_AUTOMOTIVE_SOA_EVENT_PUBLISHER_UDS_INL
+#define IOX_EXAMPLES_AUTOMOTIVE_SOA_EVENT_PUBLISHER_UDS_INL
+
+#include "owl/kom/event_publisher_uds.hpp"
+
+namespace owl
+{
+namespace kom
+{
+template <typename T>
+inline EventPublisherUds<T>::EventPublisherUds(const core::String& service,
+                                               const core::String& instance,
+                                               const core::String& event) noexcept
+    : m_publisher({service, instance, event}, {1U, iox::NodeName_t(), true})
+    , m_instanceId(instance)
+{
+}
+
+template <typename T>
+inline std::unique_ptr<T> EventPublisherUds<T>::Allocate() noexcept
+{
+    // The proxy needs some time to discover the service and create the EventSubscriberUds with the UDS server,
+    // hence the creation of the UDS client is done here
+    if (!m_calledForTheFirstTime)
+    {
+        m_uds = std::move(iox::posix::UnixDomainSocket::create(m_instanceId, iox::posix::IpcChannelSide::CLIENT)
+                              .or_else([](auto&) {
+                                  std::cout << "Failed to create UNIX domain socket!" << std::endl;
+                                  std::terminate();
+                              })
+                              .value());
+        m_calledForTheFirstTime = true;
+    }
+
+    // Allocate the memory on the heap
+    return std::make_unique<SampleType>();
+}
+
+template <typename T>
+void EventPublisherUds<T>::Send(std::unique_ptr<SampleType> userSamplePtr) noexcept
+{
+    std::string tempBuffer;
+
+    tempBuffer.append(reinterpret_cast<char*>(&userSamplePtr->counter), 4);
```

Comment:
Why write a template `EventPublisher` when you force at this position a very specific type with a `counter` member variable? I understand that you require some serialization for uds but why not restrict the `EventPublisher` for the sake of the example to trivially copyable types with a size less or equal to `iox::posix::UnixDomainSocket::MAX_MESSAGE_SIZE`?

Then your algorithm looks like:
```cxx
std::string tempBuffer(sizeof(T), '\0');
memcpy(tempBuffer.data(), userSamplePtr.get(), sizeof(T));
m_uds.send(tempBuffer).or_else([](auto&) {
        std::cerr << "Error occurred while sending on UNIX domain socket!" << std::endl;
    });
```

Otherwise this introduces a complexity which is completely unnecessary for such an example. This example is about how one could send data with iceoryx and not to present an implementation for a serializer.

And with a complex construct like a serializer one can easily introduce tons of bugs.

Reply:
I took `iceperf` as a draft with the difference of using an UNIX domain socket abstraction for `automotive_soa`.

> for the sake of the example to trivially copyable types with a size less or equal to 

The idea is to send different topic types with up to 4 MB  to be able to benchmark iceoryx vs. UNIX domain sockets (see `topic.hpp`). Hence the easier approach you described is not possible to implement as we need to fragment the data into different messages.